File: modules/admin-ui/src/main/resources/web/admin/common/js/content/site/SiteConfig.ts
module api.content.site {

    import Property = api.data.Property;
    import PropertySet = api.data.PropertySet;
    import PropertyTree = api.data.PropertyTree;
    import ApplicationKey = api.application.ApplicationKey;

    export class SiteConfig implements api.Equitable, api.Cloneable {

        private applicationKey: ApplicationKey;

        private config: PropertySet;

        constructor(builder: SiteConfigBuilder) {
            this.applicationKey = builder.applicationKey;
            this.config = builder.config;
        }

        getApplicationKey(): api.application.ApplicationKey {
            return this.applicationKey;
        }

        getConfig(): PropertySet {
            return this.config;
        }

        toJson(): Object {
            return {
                applicationKey: this.applicationKey.toString(),
                config: this.config.toJson()
            }
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, SiteConfig)) {
                return false;
            }

            var other = <SiteConfig>o;

            if (!api.ObjectHelper.equals(this.applicationKey, other.applicationKey)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.config, other.config)) {
                return false;
            }

            return true;
        }

        clone(): SiteConfig {

            return new SiteConfigBuilder(this).build();
        }

        static create(): SiteConfigBuilder {
            return new SiteConfigBuilder();
        }
    }

    export class SiteConfigBuilder {

        applicationKey: ApplicationKey;

        config: PropertySet;

        constructor(source?: SiteConfig) {
            if (source) {
                this.applicationKey = source.getApplicationKey();
                if (source.getConfig()) {
                    var newTree = new PropertyTree(source.getConfig().getTree().getIdProvider(), source.getConfig());
                    this.config = newTree.getRoot();
                }
            }
        }

        fromData(propertySet: PropertySet): SiteConfigBuilder {
            api.util.assertNotNull(propertySet, "data cannot be null");
            var applicationKey = ApplicationKey.fromString(propertySet.getString("applicationKey"));
            var siteConfig = propertySet.getPropertySet("config");
            this.setApplicationKey(applicationKey);
            this.setConfig(siteConfig);
            return this;
        }

        setApplicationKey(value: api.application.ApplicationKey): SiteConfigBuilder {
            this.applicationKey = value;
            return this;
        }

        setConfig(value: PropertySet): SiteConfigBuilder {
            this.config = value;
            return this;
        }

        build(): SiteConfig {
            return new SiteConfig(this);
        }
    }

}

File: modules/core/core-api/src/main/java/com/enonic/xp/site/SiteConfig.java
package com.enonic.xp.site;


import java.util.Objects;

import com.enonic.xp.annotation.PublicApi;
import com.enonic.xp.app.ApplicationKey;
import com.enonic.xp.data.PropertyTree;

@PublicApi
public final class SiteConfig
{
    private final ApplicationKey applicationKey;

    private final PropertyTree config;

    public SiteConfig( final Builder builder )
    {
        this.applicationKey = Objects.requireNonNull( builder.applicationKey, "applicationKey cannot be null" );
        this.config = Objects.requireNonNull( builder.config, "config cannot be null" );
    }

    public ApplicationKey getApplicationKey()
    {
        return applicationKey;
    }

    public PropertyTree getConfig()
    {
        return config;
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }

        final SiteConfig that = (SiteConfig) o;

        return Objects.equals( this.applicationKey, that.applicationKey ) && Objects.equals( this.config, that.config );
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( applicationKey, config );
    }

    public static Builder create()
    {
        return new Builder();
    }

    public static final class Builder
    {
        private ApplicationKey applicationKey;

        private PropertyTree config;

        private Builder()
        {
        }

        public Builder application( ApplicationKey value )
        {
            this.applicationKey = value;
            return this;
        }

        public Builder config( PropertyTree value )
        {
            this.config = value;
            return this;
        }

        public SiteConfig build()
        {
            return new SiteConfig( this );
        }
    }
}
